libsupport/src/ArrowInterchange.cpp

```diff
@@ -1,44 +1,160 @@
 #include "katana/ArrowInterchange.h"
 
-#include <numeric>
-
 #include "katana/Random.h"
 
 namespace {
-std::shared_ptr<arrow::ChunkedArray>
+
+katana::Result<std::shared_ptr<arrow::ChunkedArray>>
 IndexedTake(
     const std::shared_ptr<arrow::ChunkedArray>& original,
     std::shared_ptr<arrow::Array> indices) {
   // Use Take to select those indices
   arrow::Result<arrow::Datum> take_result =
       arrow::compute::Take(original, indices);
-  KATANA_LOG_ASSERT(take_result.ok());
+  if (!take_result.ok()) {
+    return KATANA_ERROR(
+        katana::ErrorCode::ArrowError,
+        "arrow builder reserve failed type: {} reason: {}",
+        original->type()->name(), take_result.status().CodeAsString());
+  }
   arrow::Datum take = std::move(take_result.ValueOrDie());
   std::shared_ptr<arrow::ChunkedArray> chunked = take.chunked_array();
   KATANA_LOG_ASSERT(chunked->num_chunks() == 1);
   return chunked;
 }
 
-}  // anonymous namespace
-
-std::shared_ptr<arrow::Array>
-katana::Unchunk(const std::shared_ptr<arrow::ChunkedArray>& original) {
+katana::Result<std::shared_ptr<arrow::Array>>
+Indices(const std::shared_ptr<arrow::ChunkedArray>& original) {
   int64_t length = original->length();
   // Build indices array, reusable across properties
   arrow::CTypeTraits<int64_t>::BuilderType builder;
   auto res = builder.Reserve(length);
-  KATANA_LOG_ASSERT(res.ok());
+  if (!res.ok()) {
+    return KATANA_ERROR(
+        katana::ErrorCode::ArrowError,
+        "arrow builder reserve failed type: {} reason: {}",
+        original->type()->name(), res.CodeAsString());
+  }
```

Comment:
@ddn0, I'm a little lost on when to print the arrow return directly and whether I should always prefer `.status().CodeAString()` to `.status()`.

Reply:
An `arrow::Status` has a code and an optional `StatusDetail`.

https://github.com/apache/arrow/blob/master/cpp/src/arrow/status.h#L306

If you pass an `arrow::Status` to `fmt`, `fmt` will attempt to use `ostream` printing which is enabled for a `Status` via https://github.com/apache/arrow/blob/master/cpp/src/arrow/status.h#L124 , and which calls `Status::ToString`

Compared to `status::CodeAsString`, `ToString` will print the optional `StatusDetail` while `CodeAsString` will just return a string version of the error code.

Reply:
Mystery resolved, thank you.  `ToStringOstreamable `is a new one for me.